implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -77,110 +76,108 @@ public KeyLocationResolver(JWTAuthContextInfo authContextInfo) throws Unresolvab
 
     @Override
     public Key resolveKey(JsonWebSignature jws, List<JsonWebStructure> nestingContext) throws UnresolvableKeyException {
-        if (verificationKey != null) {
-            return verificationKey;
-        }
-        PublicKey key = tryAsJWK(jws);
-        if (key == null) {
-            key = tryAsPEM();
-        }
-        if (key == null) {
-            throw new UnresolvableKeyException("Failed to read location as any of JWK, JWKS, PEM. "
-                    + authContextInfo.getPublicKeyLocation());
-        }
-        return key;
-    }
-
-    private PublicKey tryAsPEM() {
-        try {
-            verificationKey = KeyUtils.decodePublicKey(content);
-        } catch (Exception e) {
-            log.debug("Failed to read location as PEM", e);
+        // httpsJwks may refresh itself even if the jwksRefreshInterval is set to 0 as it checks HTTPS Cache-Control header
+        // so the public key will have to be created per request to ensure the rotation (if any) is effective.
+        if (httpsJwks != null && authContextInfo.getTokenKeyId() == null)
+            return getHttpsJwk(httpsJwks, jws);
+
+        if (verificationKey == null) {
+            synchronized (this) {
+                if (verificationKey == null) {
+                    PublicKey localKey = tryAsLocalJwk(content, jws);
+
+                    if (localKey == null) {
+                        localKey = tryAsPEM(content);
+                    }
+                    if (localKey == null) {
+                        throw new UnresolvableKeyException(FAILED_TO_READ_LOCATION_MESSAGE
+                                + authContextInfo.getPublicKeyLocation());
+                    }
+                    verificationKey = localKey;
```

Comment:
Oh well, after looking at this code for a while, I've realized the new code has changed the original behaviour and introduced a bug. Basically, if we work with the local JWK sets we can not assign the 'localKey' to the `verificationKey` unless this (old) [condition](https://github.com/smallrye/smallrye-jwt/blob/master/implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java#L128) is true: this is because JWK set may have many JWKs, each of them with its own `kid`, so unless that condition is true, and more specifically, `authContextInfo.getTokenKeyId() != null` is true (because if this property is set it means the users wants to deal with a specific `kid` only and hence the fact JWK set has many keys does not matter), we can not initialize `verificationKey`.
There should've been a test, but well, I've spotted it now.
I honestly don't know now what are we doing with this KeyLocationResolver code.
`verificationKey` can not be not null in 2 cases:
1. tryAsPEM - PEM key file has no KID to match against
2. JWK set and authContextInfo.getTokenId() != null
And it can't be both 1 and 2 for a given KeyLocationResolver instance.

So why can't we return to the drawing board again and see what are we trying to protect against, as far as the thread safety is concerned, in `KeyLocationResolver` ? Explain me please, why just having `volatile` is not enough given what I have just explained.

I think `getKid(jws)` is definitely the good thing that came out of our discussions.
The other thing which I've realized is that tryAsPEM can be done from the constructor immediately after loadContent().